### PR TITLE
TW-1058 Added initial and additional loaders

### PR DESCRIPTION
**Pull request by herkoss**

### CODE REVIEW

File: src/components/activity-groups-list/activity-groups-list.tsx [hunk 1]
```diff
@@ -71,32 +84,31 @@ export const ActivityGroupsList: FC<Props> = ({
...
-        refreshControl={<RefreshControl {...fakeRefreshControlProps} />}
-      />
-    </>
+    <SectionList
+      sections={sections}
+      stickySectionHeadersEnabled={true}
+      contentContainerStyle={styles.sectionListContentContainer}
+      onEndReachedThreshold={0.01}
+      onEndReached={handleUpdate}
+      keyExtractor={item => item[0].hash}
```
Comment (lendihop) [inline src/components/activity-groups-list/activity-groups-list.tsx:93]: extract to a separate fn and move outside component

File: src/components/activity-groups-list/activity-groups-list.tsx [hunk 2]
```diff
@@ -71,32 +84,31 @@ export const ActivityGroupsList: FC<Props> = ({
...
+      renderItem={({ item, index, section }) => (
```
Comment (lendihop) [inline src/components/activity-groups-list/activity-groups-list.tsx:94]: extract to a separate fn

File: src/hooks/use-contract-activity.ts [hunk 3]
```diff
@@ -15,28 +15,37 @@ export const useContractActivity = (tokenSlug?: string): UseActivityInterface =>
...
-        });
-      } else {
-        setActivities(activities);
+        if (activities.length === 0) {
+          setIsAllLoaded(true);
+        }
+        if (refresh === true) {
+          setActivities(prev => {
+            const allActivities = [...activities, ...prev];
+            const allHashes = allActivities.map(x => x[0]).map(x => x.hash);
```
Comment (svjatoslavtt) [inline src/hooks/use-contract-activity.ts:34]: Please, avoid the short name in functions `v: value, n: number etc.`

File: src/hooks/use-contract-activity.ts [hunk 4]
```diff
@@ -15,28 +15,37 @@ export const useContractActivity = (tokenSlug?: string): UseActivityInterface =>
...
+            const onlyUniqueHashes = uniq(allHashes);
+            const onlyUniqueActivitites = onlyUniqueHashes
+              .map(x => allActivities.find(y => y[0].hash === x))
```
Comment (svjatoslavtt) [inline src/hooks/use-contract-activity.ts:37]: the same

**Review by lendihop — DISMISSED** [review]

File: src/components/activity-groups-list/activity-groups-list.tsx (hunk 2 above)
Comment (Boltman92) [inline src/components/activity-groups-list/activity-groups-list.tsx:94]: move `renderItem` to const please

**Review by Boltman92 — DISMISSED** [review]

File: ios/TempleWallet.xcodeproj/project.pbxproj
Comment (lendihop) [inline ios/TempleWallet.xcodeproj/project.pbxproj:1]: pls revert this changes

File: src/hooks/use-contract-activity.ts [hunk 5]
```diff
@@ -51,31 +62,48 @@ export const useContractActivity = (tokenSlug?: string): UseActivityInterface =>
...
+              setActivities(prev => [...prev, ...newActivity]);
             }
-            setActivities(prev => [...prev, ...newActivity]);
           }
         }
       }
+    } catch (error) {
+      console.error(error);
+    } finally {
+      await sleep(1000);
```
Comment (lendihop) [inline src/hooks/use-contract-activity.ts:91]: what is this for?

File: ios/TempleWallet.xcodeproj/project.pbxproj
Comment (lendihop) [inline ios/TempleWallet.xcodeproj/project.pbxproj:1]: pls revert all changes in this file

File: src/hooks/use-contract-activity.ts (hunk 5 above)
Comment (lendihop) [reply src/hooks/use-contract-activity.ts:91]: did you test it on different devices?

Comment (herkoss) [reply src/hooks/use-contract-activity.ts:91]: Only on my iphone and emulator

**Review by lendihop — DISMISSED** [review]

**Review by Boltman92 — APPROVED** [review]

File: src/hooks/use-contract-activity.ts [hunk 6]
```diff
@@ -15,28 +15,37 @@ export const useContractActivity = (tokenSlug?: string): UseActivityInterface =>
   const lastActivityRef = useRef<string>('');
 
   const [isAllLoaded, setIsAllLoaded] = useState<boolean>(false);
+  const [isInitialLoading, setIsInitialLoading] = useState<boolean>(false);
+  const [isAdditionalLoading, setIsAdditionalLoading] = useState<boolean>(false);
```
[the remark this answers is not in the document]
Comment (keshan3262) [reply src/hooks/use-contract-activity.ts:19]: I think it should be enough. I just was too lazy to review the changes before mine.

**Review by lendihop — DISMISSED** [review]

**Review by lendihop — DISMISSED** [review]

**Review by lendihop — APPROVED** [review]

**Review by lendihop — APPROVED** [review]

File: src/hooks/use-contract-activity.ts [hunk 7]
```diff
@@ -51,31 +68,54 @@ export const useContractActivity = (tokenSlug?: string): UseActivityInterface =>
...
         }
       }
+    } catch (error) {
+      console.error(error);
+    } finally {
+      if (wasLoading) {
+        return;
+      }
+
+      const isAllLoaded = newActivities.length === 0 && !wasLoading;
```
Comment (lendihop) [inline src/hooks/use-contract-activity.ts:98]: so the new activities chunk will be loaded but not set, mb there is no need to load it?

File: src/hooks/use-contract-activity.ts [hunk 8]
```diff
@@ -51,31 +68,54 @@ export const useContractActivity = (tokenSlug?: string): UseActivityInterface =>
   };
 
   const handleUpdate = async () => {
-    if (activities.length > 0 && !isAllLoaded) {
-      const lastActivityGroup = activities[activities.length - 1].sort((a, b) => b.id - a.id);
+    let newActivities: Array<ActivityGroup> = [];
+    const wasLoading = isLoading;
```
Comment (lendihop) [inline src/hooks/use-contract-activity.ts:72]: mb `if (isLoading) return;` ?

Comment (keshan3262) [reply src/hooks/use-contract-activity.ts:72]: I think it's not the case, we may need to update `lastActivityRef` value

File: src/hooks/use-contract-activity.ts (hunk 7 above)
Comment (lendihop) [reply src/hooks/use-contract-activity.ts:98]: ok, sorry, i didnt notice this line

**Review by lendihop — APPROVED** [review]

**Review by NovakVlad2424 — APPROVED** [review]
ios done: 1.16_1031
